### PR TITLE
Refactor some utility Linq methods into a new file

### PR DESCRIPTION
**Pull request by jamesqo**
There are some helper methods in the Linq codebase that are used by different files, e.g. `CombineSelectors` is used by Select.cs and Where.cs, but is put in Select.cs. This PR just moves a couple of them into Common files to make it less confusing.

Changes:

- `CombinePredicates` and `CombineSelectors` have been moved to `FunctionUtilities`

- `EquivalentEqualityComparers` has been renamed to `AreEqualityComparersEqual` and put in `ComparerUtilities`. I also added a new method, `AreComparersEqual`.

  - I had planned to use these methods in `HashSet` and `SortedSet` as well since they have similar utility methods, but realized that their methods [accept a set](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/HashSet.cs#L1675) rather than a comparer, so I'm leaving them alone for now.

cc @stephentoub

### CODE REVIEW

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            }
+
+            var @default = EqualityComparer<T>.Default;
+
+            if (left == null)
+            {
+                // Micro-opt: Typically it's impossible to get a different instance
+                // of the default comparer without reflection/serialization.
+                // Save a virtual method call to Equals in the common case.
+                return right == @default || right.Equals(@default);
```
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: I changed the method slightly here by doing a ref-equality check first before calling `Equals`; see the comment.

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    internal static class ComparerUtilities
+    {
+        public static bool AreComparersEqual<T>(IComparer<T> left, IComparer<T> right)
+        {
+            if (left == right)
+            {
+                return true;
+            }
+
+            var @default = Comparer<T>.Default;
```
Comment (AlexGhiondea) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: nit: rename this to defaultComparer. It removes the need to have @ in the variable name.

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        public static bool AreEqualityComparersEqual<T>(IEqualityComparer<T> left, IEqualityComparer<T> right)
+        {
+            if (left == right)
+            {
+                return true;
+            }
+
+            var @default = EqualityComparer<T>.Default;
```
Comment (AlexGhiondea) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: Same comment as above.

File: src/Common/src/System/Collections/Generic/FunctionUtilities.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+namespace System.Collections.Generic
+{
+    internal static class FunctionUtilities
+    {
+        public static Func<TSource, bool> CombinePredicates<TSource>(Func<TSource, bool> predicate1, Func<TSource, bool> predicate2)
+        {
+            return x => predicate1(x) && predicate2(x);
+        }
+
+        public static Func<TSource, TResult> CombineSelectors<TSource, TMiddle, TResult>(Func<TSource, TMiddle> selector1, Func<TMiddle, TResult> selector2)
```
Comment (AlexGhiondea) [inline src/Common/src/System/Collections/Generic/FunctionUtilities.cs]: nit: consider renaming TMiddle to TIntermediate.

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Collections.Generic
+{
+    internal static class ComparerUtilities
+    {
+        public static bool AreComparersEqual<T>(IComparer<T> left, IComparer<T> right)
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: Where is this used?  Let's not add it unless it's actually needed.

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs (hunk 2 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: Done.

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs (hunk 3 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: Done.

File: src/Common/src/System/Collections/Generic/FunctionUtilities.cs (hunk 4 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/FunctionUtilities.cs]: I think TMiddle is more concise than TIntermediate here...

File: src/Common/src/System/Collections/Generic/ComparerUtilities.cs (hunk 5 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/ComparerUtilities.cs]: Removed.

**Comment by jamesqo** [comment]
@stephentoub I decided to go in a different direction with this PR-- instead of a Common file, the utility functions are now in `System\Linq\Utilities.cs`. Additionally the `using static` at the top of the file becomes much shorter, I decided to keep it though since 'utilities' is somewhat of a pain to type.

**Comment by stephentoub** [comment]
@jamesqo, can you please rebase this to fix the conflict?

**Comment by jamesqo** [comment]
@stephentoub Done. Thanks for reviewing.

**Pull request merged**